src/Features/Core/Portable/Workspace/BackgroundParser.cs

```diff
@@ -178,8 +178,12 @@ private void ParseDocumentAsync(Document document)
 
             var cancellationToken = cancellationTokenSource.Token;
 
+            // once task runs, we never cancel running parsing request. this will make sure 
+            // chained incremental parsing request will always get shorten by BG
+            //
+            // we are alraedy running in BG, you no need to run async version
```

Comment:
I'd expand on this comment a bit.  Maybe something like:

> We end up creating a chain of parsing tasks that each attempt to produce the appropriate syntax tree for any given document.   Once we start work to create the syntax tree for a given document, we don't want to stop.  Otherwise we can end up in the unfortunate scenario where we keep cancelling work, and then having the next task *re-do* the work we were just in the middle of.  By not cancelling, we can reuse the *useful* results of previous tasks when performing later steps in the chain.

Reply:
Note: this PR has a good description of the issue.  IMO, it would be valuable to put that description into the actual code.